base_exception/tests/test_base_exception.py

```diff
@@ -2,115 +2,107 @@
-                "code": "if self.amount_total <= 200.0: failed=True",
+                "code": "if not self.partner_id.zip: failed=True",
+                "exception_type": "by_py_code",
             }
         )
 
-        cls.exceptionno_lineqty = cls.env["exception.rule"].create(
+    @classmethod
+    def tearDownClass(cls):
+        cls.loader.restore_registry()
+        super().tearDownClass()
+
+    def test_valid(self):
+        self.exception_rule.active = False
+        self.po.button_confirm()
+        self.assertFalse(self.po.exception_ids)
+
+    def test_fail_by_py(self):
+        with self.assertRaises(ValidationError):
+            self.po.button_confirm()
+        self.assertTrue(self.po.exception_ids)
+
+    def test_fail_by_domain(self):
+        self.exception_rule.write(
             {
-                "name": "Qty > 0",
-                "sequence": 10,
-                "model": "base.exception.test.purchase.line",
-                "code": "if obj.qty <= 0: failed=True",
+                "domain": "[('partner_id.zip', '=', False)]",
+                "exception_type": "by_domain",
             }
         )
+        with self.assertRaises(ValidationError):
+            self.po.button_confirm()
+        self.assertTrue(self.po.exception_ids)
 
-    def test_purchase_order_exception(self):
-        partner = self.env.ref("base.res_partner_1")
-        partner.zip = False
-        potest1 = self.env["base.exception.test.purchase"].create(
+    def test_fail_by_method(self):
+        self.exception_rule.write(
             {
-                "name": "Test base exception to basic purchase",
-                "partner_id": partner.id,
-                "line_ids": [
-                    (0, 0, {"name": "line test", "amount": 120.0, "qty": 1.5})
-                ],
+                "method": "exception_method_no_zip",
+                "exception_type": "by_method",
             }
         )
+        with self.assertRaises(ValidationError):
+            self.po.button_confirm()
+        self.assertTrue(self.po.exception_ids)
+
+    def test_ignorable_exception(self):
         # Block because of exception during validation
         with self.assertRaises(ValidationError):
-            potest1.button_confirm()
+            self.po.button_confirm()
         # Test that we have linked exceptions
-        self.assertTrue(potest1.exception_ids)
+        self.assertTrue(self.po.exception_ids)
         # Test ignore exeception make possible for the po to validate
-        potest1.action_ignore_exceptions()
-        self.assertTrue(potest1.ignore_exception)
-        self.assertFalse(potest1.exceptions_summary)
-        potest1.button_confirm()
-        self.assertTrue(potest1.state == "purchase")
+        self.po.action_ignore_exceptions()
+        self.assertTrue(self.po.ignore_exception)
+        self.assertFalse(self.po.exceptions_summary)
+        self.po.button_confirm()
+        self.assertTrue(self.po.state == "purchase")
```

Comment:
not blocking, just for information but `assert self.po.state == "purchase"` is a more pythonic syntax

Reply:
I didn't see it, thx. I replace it with : self.assertEqual(self.po.state, "purchase")

Reply:
It's not the same !

Reply:
@hparfr you talk about 

 ```assert self.po.state == "purchase" ``` (this is indeed not valid)

or

 ```self.assertEqual(self.po.state, "purchase") ```

Reply:
Why not valid

Reply:
David says you can use the `assert` keyword instead of `self.assertSomething()` method.
It's one of the key arguments of pytest: 

> Due to pytest’s detailed assertion introspection, only plain assert statements are used
> [...]    Detailed info on failing assert statements (no need to remember self.assert* names)
https://docs.pytest.org/en/7.0.x/index.html

Reply:
And yes, `assertEquals`, in this case is more expressive than `assertTrue`, but it was not the point here :dart: .

Reply:
https://realpython.com/python-assert-statement/#testing-your-code-with-assertions